torch/csrc/jit/passes/pattern_fusion.cpp

```diff
@@ -0,0 +1,134 @@
+#include <torch/csrc/jit/passes/pattern_fusion.h>
+#include <torch/csrc/jit/irparser.h>
+#include <torch/csrc/jit/subgraph_matcher.h>
+
+namespace torch {
+namespace jit {
+
+void PatternFuser::RegisterDefaultPatterns() {
+  // TODO: Add actual patterns (like Conv-Relu).
+  RegisterFusionPattern(
+      R"IR(
+graph(%x, %w, %b):
+  %c = aten::conv(%x, %w, %b)
+  %r = aten::relu(%c)
+  return (%r))IR",
+      /*fused_node_name=*/"aten::convrelu",
+      /*inputs=*/{"x", "w", "b"},
+      /*outputs=*/{"r"});
+}
+
+void PatternFuser::RegisterFusionPattern(
+    const std::string& pattern,
+    const std::string& fused_node_name,
+    std::vector<std::string> inputs,
+    std::vector<std::string> outputs) {
+  FusionPatternDescr d = {pattern, fused_node_name, inputs, outputs};
+  patterns_.push_back(d);
+}
```

Comment:
We need to handle cases where patterns overlap and order them. For example: consider two fusion patterns: q->conv->sum->relu->dq, which can be replaced with a qconvsumrelu and a smaller pattern: sum->relu, which gets replaced with a sumrelu. We want the second fusion to happen only when the first fusion cannot occur. We will need to order the pattern matching from the largest to the smallest patterns. This is easier when we have patterns that are proper subsets of each other, as in the example below. Otherwise, we could go with the length of the pattern/some other heuristic to order the pattern matches.

Reply:
This is controlled by the order in which the patterns are registered. In the case you mentioned, we will just register the big pattern first.